db/db_impl/db_impl.cc

```diff
@@ -1828,126 +1764,350 @@ std::vector<Status> DBImpl::MultiGet(
+  PERF_TIMER_STOP(get_snapshot_time);
+
+  return last_try;
+}
+
+void DBImpl::MultiGet(const ReadOptions& read_options, const size_t num_keys,
+                      ColumnFamilyHandle** column_families, const Slice* keys,
+                      PinnableSlice* values, Status* statuses,
+                      const bool sorted_input) {
+  autovector<KeyContext, MultiGetContext::MAX_BATCH_SIZE> key_context;
+  autovector<KeyContext*, MultiGetContext::MAX_BATCH_SIZE> sorted_keys;
+  sorted_keys.resize(num_keys);
+  for (size_t i = 0; i < num_keys; ++i) {
+    key_context.emplace_back(column_families[i], keys[i], &values[i],
+                             &statuses[i]);
+  }
+  for (size_t i = 0; i < num_keys; ++i) {
+    sorted_keys[i] = &key_context[i];
+  }
+  PrepareMultiGetKeys(num_keys, sorted_input, &sorted_keys);
+
+  struct MultiGetColumnFamilyData {
+    ColumnFamilyHandle* cf;
+    ColumnFamilyData* cfd;
+    size_t start;
+    size_t num_keys;
+    SuperVersion* super_version;
+    MultiGetColumnFamilyData(ColumnFamilyHandle* column_family, size_t first,
+                             size_t count, SuperVersion* sv)
+        : cf(column_family), start(first), num_keys(count), super_version(sv) {
+      ColumnFamilyHandleImpl* cfh =
+          reinterpret_cast<ColumnFamilyHandleImpl*>(cf);
+      cfd = cfh->cfd();
+    }
+    MultiGetColumnFamilyData() = default;
+  };
+  autovector<MultiGetColumnFamilyData, MultiGetContext::MAX_BATCH_SIZE>
+      multiget_cf_data;
+  size_t cf_start = 0;
+  ColumnFamilyHandle* cf = sorted_keys[0]->column_family;
+  for (size_t i = 0; i < num_keys; ++i) {
+    KeyContext* key_ctx = sorted_keys[i];
+    if (key_ctx->column_family != cf) {
+      multiget_cf_data.emplace_back(
+          MultiGetColumnFamilyData(cf, cf_start, i - cf_start, nullptr));
+      cf_start = i;
+      cf = key_ctx->column_family;
+    }
+  }
+  {
+    multiget_cf_data.emplace_back(
+        MultiGetColumnFamilyData(cf, cf_start, num_keys - cf_start, nullptr));
+  }
+  struct IterDerefFunc {
+    inline MultiGetColumnFamilyData* operator()(
+        autovector<MultiGetColumnFamilyData,
+                   MultiGetContext::MAX_BATCH_SIZE>::iterator& cf_iter) {
+      return &(*cf_iter);
+    }
+  };
+
+  SequenceNumber snapshot;
+  bool unref_only = MultiCFSnapshot<
+      autovector<MultiGetColumnFamilyData, MultiGetContext::MAX_BATCH_SIZE>,
+      IterDerefFunc>(read_options, nullptr, &multiget_cf_data, &snapshot);
+
+  for (auto cf_iter = multiget_cf_data.begin();
+       cf_iter != multiget_cf_data.end(); ++cf_iter) {
+    MultiGetImpl(read_options, cf_iter->start, cf_iter->num_keys, &sorted_keys,
+                 cf_iter->super_version, snapshot, nullptr, nullptr);
+    if (!unref_only) {
+      ReturnAndCleanupSuperVersion(cf_iter->cfd, cf_iter->super_version);
+    } else {
+      cf_iter->cfd->GetSuperVersion()->Unref();
+    }
+  }
+}
+
 // Order keys by CF ID, followed by key contents
 struct CompareKeyContext {
```

Comment:
It's not related to this PR but I'm a little bit confused by this struct. Is it something that can be hidden in an anonymous namespace?

Reply:
Yes, it can be in an anonymous namespace.